Compilation failure on Ubuntu 22.04 with OpenMPI on recent commit

The compilation of the development branch seems to fail on Ubuntu 22.04 with OpenMPI after the most recent commits with the following error:
```
107.2 [ 51%] Linking CXX executable t_relay_mpi_smoke
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Op::Init(void (*)(void const*, void*, int, MPI::Datatype const&), bool)':
107.3 t_relay_mpi_smoke.cpp:(.text._ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb[_ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb]+0x1d): undefined reference to `ompi_mpi_cxx_op_intercept'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Intracomm::Clone() const':
107.3 t_relay_mpi_smoke.cpp:(.text._ZNK3MPI9Intracomm5CloneEv[_ZNK3MPI9Intracomm5CloneEv]+0x40): undefined reference to `MPI::Comm::Comm()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Graphcomm::Clone() const':
107.3 t_relay_mpi_smoke.cpp:(.text._ZNK3MPI9Graphcomm5CloneEv[_ZNK3MPI9Graphcomm5CloneEv]+0x3a): undefined reference to `MPI::Comm::Comm()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Cartcomm::Sub(bool const*) const':
107.3 t_relay_mpi_smoke.cpp:(.text._ZNK3MPI8Cartcomm3SubEPKb[_ZNK3MPI8Cartcomm3SubEPKb]+0x96): undefined reference to `MPI::Comm::Comm()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Intracomm::Create_graph(int, int const*, int const*, bool) const':
107.3 t_relay_mpi_smoke.cpp:(.text._ZNK3MPI9Intracomm12Create_graphEiPKiS2_b[_ZNK3MPI9Intracomm12Create_graphEiPKiS2_b]+0x42): undefined reference to `MPI::Comm::Comm()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o: in function `MPI::Cartcomm::Clone() const':
107.3 t_relay_mpi_smoke.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x3a): undefined reference to `MPI::Comm::Comm()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o:t_relay_mpi_smoke.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0xab): more undefined references to `MPI::Comm::Comm()' follow
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o:(.data.rel.ro._ZTVN3MPI8DatatypeE[_ZTVN3MPI8DatatypeE]+0x78): undefined reference to `MPI::Datatype::Free()'
107.3 /usr/bin/ld: CMakeFiles/t_relay_mpi_smoke.dir/t_relay_mpi_smoke.cpp.o:(.data.rel.ro._ZTVN3MPI3WinE[_ZTVN3MPI3WinE]+0x48): undefined reference to `MPI::Win::Free()'
107.3 collect2: error: ld returned 1 exit status
107.3 make[2]: *** [tests/relay/CMakeFiles/t_relay_mpi_smoke.dir/build.make:102: tests/relay/t_relay_mpi_smoke] Error 1
107.3 make[1]: *** [CMakeFiles/Makefile2:3799: tests/relay/CMakeFiles/t_relay_mpi_smoke.dir/all] Error 2
107.3 make[1]: *** Waiting for unfinished jobs....
109.1 [ 52%] Linking CXX executable t_relay_mpi_test
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Op::Init(void (*)(void const*, void*, int, MPI::Datatype const&), bool)':
109.2 t_relay_mpi_test.cpp:(.text._ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb[_ZN3MPI2Op4InitEPFvPKvPviRKNS_8DatatypeEEb]+0x1d): undefined reference to `ompi_mpi_cxx_op_intercept'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Intracomm::Clone() const':
109.2 t_relay_mpi_test.cpp:(.text._ZNK3MPI9Intracomm5CloneEv[_ZNK3MPI9Intracomm5CloneEv]+0x40): undefined reference to `MPI::Comm::Comm()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Graphcomm::Clone() const':
109.2 t_relay_mpi_test.cpp:(.text._ZNK3MPI9Graphcomm5CloneEv[_ZNK3MPI9Graphcomm5CloneEv]+0x3a): undefined reference to `MPI::Comm::Comm()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Cartcomm::Sub(bool const*) const':
109.2 t_relay_mpi_test.cpp:(.text._ZNK3MPI8Cartcomm3SubEPKb[_ZNK3MPI8Cartcomm3SubEPKb]+0x96): undefined reference to `MPI::Comm::Comm()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Intracomm::Create_graph(int, int const*, int const*, bool) const':
109.2 t_relay_mpi_test.cpp:(.text._ZNK3MPI9Intracomm12Create_graphEiPKiS2_b[_ZNK3MPI9Intracomm12Create_graphEiPKiS2_b]+0x42): undefined reference to `MPI::Comm::Comm()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o: in function `MPI::Cartcomm::Clone() const':
109.2 t_relay_mpi_test.cpp:(.text._ZNK3MPI8Cartcomm5CloneEv[_ZNK3MPI8Cartcomm5CloneEv]+0x3a): undefined reference to `MPI::Comm::Comm()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o:t_relay_mpi_test.cpp:(.text._ZNK3MPI9Intracomm11Create_cartEiPKiPKbb[_ZNK3MPI9Intracomm11Create_cartEiPKiPKbb]+0xab): more undefined references to `MPI::Comm::Comm()' follow
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o:(.data.rel.ro._ZTVN3MPI8DatatypeE[_ZTVN3MPI8DatatypeE]+0x78): undefined reference to `MPI::Datatype::Free()'
109.2 /usr/bin/ld: CMakeFiles/t_relay_mpi_test.dir/t_relay_mpi_test.cpp.o:(.data.rel.ro._ZTVN3MPI3WinE[_ZTVN3MPI3WinE]+0x48): undefined reference to `MPI::Win::Free()'
109.2 collect2: error: ld returned 1 exit status
```
To reproduce, please find attached two Dockerfiles. 
[dockerfiles.zip](https://github.com/user-attachments/files/16317527/dockerfiles.zip)

To build the faulty container:
```
docker build --no-cache -f ./Dockerfile_current -t conduit-ubuntu:current .
``` 
To build the stable container:
```
docker build --no-cache -f ./Dockerfile_stable -t conduit-ubuntu:stable .
```

--- a/src/cmake/SetupBLT.cmake
+++ b/src/cmake/SetupBLT.cmake
@@ -39,6 +39,16 @@ if(ENABLE_MPI)
         set(MPI_FOUND ON CACHE BOOL "")
     endif()
 
+    #################
+    # defines to prevent mpi from using C++ apis and introducing link
+    #  dep of libmpi_cxx
+    # (encountred when using openmpi as mpi)
+    #
+    # In theory, calling set(MPI_CXX_SKIP_MPICXX TRUE BOOL) before FindMPI
+    # would also add these flags, but that did not work w/ cmake 3.22
+    #################
+    set(CONDUIT_MPI_SKIP_MPICXX_DEFINES MPICH_SKIP_MPICXX OMPI_SKIP_MPICXX _MPICC_H)
+
     # if we are using BLT's enable mpi, then we must
     # make sure the MPI targets exist
     if(ENABLE_FIND_MPI)

--- a/src/cmake/SetupBLT.cmake
+++ b/src/cmake/SetupBLT.cmake
@@ -41,8 +41,7 @@ if(ENABLE_MPI)
 
     #################
     # defines to prevent mpi from using C++ apis and introducing link
-    #  dep of libmpi_cxx
-    # (encountred when using openmpi as mpi)
+    # dep of libmpi_cxx  (encountred when using openmpi as mpi)
     #
     # In theory, calling set(MPI_CXX_SKIP_MPICXX TRUE BOOL) before FindMPI
     # would also add these flags, but that did not work w/ cmake 3.22

--- a/src/config/conduit_config.mk.in
+++ b/src/config/conduit_config.mk.in
@@ -65,6 +65,13 @@ CONDUIT_LINK_RPATH += $(if $(CONDUIT_PARMETIS_DIR), $(CONDUIT_PARMETIS_RPATH_FLA
 CONDUIT_LINK_RPATH += $(if $(CONDUIT_CALIPER_DIR), $(CONDUIT_CALIPER_RPATH_FLAGS_VALUE))
 CONDUIT_LINK_RPATH += $(if $(CONDUIT_ADIAK_DIR), $(CONDUIT_ADIAK_RPATH_FLAGS_VALUE))
 
+#################
+# Defines
+#################
+
+# defines to prevent mpi from using C++ apis and introducing link
+# dep of libmpi_cxx  (encountred when using openmpi as mpi)
+CONDUIT_MPI_DEFINES = -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -D_MPICC_H
 
 #################
 # Include Flags
@@ -85,6 +92,8 @@ CONDUIT_INCLUDE_FLAGS += $(if $(CONDUIT_HDF5_DIR),$(CONDUIT_HDF5_EXTRA_INCLUDE_F
 CONDUIT_INCLUDE_FLAGS += $(if $(CONDUIT_ZFP_DIR),-I$(CONDUIT_ZFP_DIR)/include)
 
 CONDUIT_INCLUDE_FLAGS += $(if $(CONDUIT_CALIPER_DIR),-I$(CONDUIT_CALIPER_DIR)/include)
+
+CONDUIT_INCLUDE_FLAGS += $(CONDUIT_MPI_DEFINES)
 
 # Note: parmetis headers are not exposed in our api
 

--- a/src/libs/blueprint/CMakeLists.txt
+++ b/src/libs/blueprint/CMakeLists.txt
@@ -240,6 +240,8 @@ add_compiled_library(NAME conduit_blueprint_mpi
 
 blt_add_target_compile_flags(TO conduit_blueprint_mpi FLAGS "-DCONDUIT_BLUEPRINT_MPI_ENABLED")
 
+# avoid mpi C++ apis + link dep of libmpi_cxx
+target_compile_definitions(conduit_blueprint_mpi PUBLIC ${CONDUIT_MPI_SKIP_MPICXX_DEFINES} )
 
 endif() # end if MPI_FOUND
 

--- a/src/libs/relay/CMakeLists.txt
+++ b/src/libs/relay/CMakeLists.txt
@@ -336,6 +336,9 @@ add_compiled_library(NAME conduit_relay_mpi
                      HEADERS_DEST_DIR include/conduit
                      FOLDER libs)
 
+# avoid mpi C++ apis + link dep of libmpi_cxx
+target_compile_definitions(conduit_relay_mpi PUBLIC ${CONDUIT_MPI_SKIP_MPICXX_DEFINES} )
+
 
 ################################################################
 # If we have mpi, add the conduit relay mpi io library
@@ -412,5 +415,9 @@ add_compiled_library(NAME conduit_relay_mpi_io
                      FOLDER libs)
 
 blt_add_target_compile_flags(TO conduit_relay_mpi_io FLAGS "-DCONDUIT_RELAY_IO_MPI_ENABLED")
+
+# avoid mpi C++ apis + link dep of libmpi_cxx
+target_compile_definitions(conduit_relay_mpi_io PUBLIC ${CONDUIT_MPI_SKIP_MPICXX_DEFINES} )
+
 
 endif() # end if MPI_FOUND

